### PR TITLE
Replace excavate with YARA, and introduce web_parameters

### LINKED ISSUE #1252: New Excavate Module: Implement Yara
There appear to be some slow regexes in excavate:

```
[WARN] Executing <Task pending name='excavate._worker()' coro=<BaseModule._worker() running at /home/bls/Downloads/code/bbot/bbot/modules/base.py:629> wait_for=<Task pending name='excavate.handle_event(HTTP_RESPONSE("{\'url\': \'https://shop.tesla.com/category/vehicle-accessories\', \'timestamp\': \'202...", module=httpx, tags={\'cdn-akamai\', \'status-200\', \'ip-23-203-244-93\', \'http-title-tesla-vehicle-accessories\', \'endpoint\', \'in-scope\'}))' coro=<excavate.handle_event() running at /home/bls/Downloads/code/bbot/bbot/modules/internal/excavate.py:364> cb=[Task.task_wakeup()] created at /usr/lib/python3.11/asyncio/tasks.py:384> created at /usr/lib/python3.11/asyncio/tasks.py:384> took 0.103 seconds
[WARN] Executing <Task pending name='excavate.handle_event(HTTP_RESPONSE("{\'url\': \'https://shop.tesla.com/category/vehicle-accessories\', \'timestamp\': \'202...", module=httpx, tags={\'cdn-akamai\', \'status-200\', \'ip-23-203-244-93\', \'http-title-tesla-vehicle-accessories\', \'endpoint\', \'in-scope\'}))' coro=<excavate.handle_event() running at /home/bls/Downloads/code/bbot/bbot/modules/internal/excavate.py:392> cb=[Task.task_wakeup()] created at /usr/lib/python3.11/asyncio/tasks.py:384> took 9.984 seconds
[WARN] Executing <Task pending name='excavate.handle_event(HTTP_RESPONSE("{\'url\': \'https://shop.tesla.com/category/vehicle-accessories\', \'timestamp\': \'202...", module=httpx, tags={\'cdn-akamai\', \'status-200\', \'ip-23-203-244-93\', \'http-title-tesla-vehicle-accessories\', \'endpoint\', \'in-scope\'}))' coro=<excavate.handle_event() running at /home/bls/Downloads/code/bbot/bbot/modules/internal/excavate.py:392> cb=[Task.task_wakeup()] created at /usr/lib/python3.11/asyncio/tasks.py:384> took 0.654 seconds
[WARN] Executing <Task pending name='excavate.handle_event(HTTP_RESPONSE("{\'url\': \'https://shop.tesla.com/category/vehicle-accessories\', \'timestamp\': \'202...", module=httpx, tags={\'cdn-akamai\', \'status-200\', \'ip-23-203-244-93\', \'http-title-tesla-vehicle-accessories\', \'endpoint\', \'in-scope\'}))' coro=<excavate.handle_event() running at /home/bls/Downloads/code/bbot/bbot/modules/internal/excavate.py:392> cb=[Task.task_wakeup()] created at /usr/lib/python3.11/asyncio/tasks.py:384> took 0.443 seconds
[WARN] Executing <Task pending name='excavate.handle_event(HTTP_RESPONSE("{\'url\': \'https://shop.tesla.com/category/vehicle-accessories\', \'timestamp\': \'202...", module=httpx, tags={\'cdn-akamai\', \'status-200\', \'ip-23-203-244-93\', \'http-title-tesla-vehicle-accessories\', \'endpoint\', \'in-scope\'}))' coro=<excavate.handle_event() running at /home/bls/Downloads/code/bbot/bbot/modules/internal/excavate.py:392> cb=[Task.task_wakeup()] created at /usr/lib/python3.11/asyncio/tasks.py:384> took 0.154 seconds
[WARN] Executing <Task pending name='excavate.handle_event(HTTP_RESPONSE("{\'url\': \'https://shop.tesla.com/category/vehicle-accessories\', \'timestamp\': \'202...", module=httpx, tags={\'cdn-akamai\', \'status-200\', \'ip-23-203-244-93\', \'http-title-tesla-vehicle-accessories\', \'endpoint\', \'in-scope\'}))' coro=<excavate.handle_event() running at /home/bls/Downloads/code/bbot/bbot/modules/internal/excavate.py:392> cb=[Task.task_wakeup()] created at /usr/lib/python3.11/asyncio/tasks.py:384> took 1.628 seconds
[WARN] Executing <Task pending name='excavate.handle_event(HTTP_RESPONSE("{\'url\': \'https://shop.tesla.com/category/vehicle-accessories\', \'timestamp\': \'202...", module=httpx, tags={\'cdn-akamai\', \'status-200\', \'ip-23-203-244-93\', \'http-title-tesla-vehicle-accessories\', \'endpoint\', \'in-scope\'}))' coro=<excavate.handle_event() running at /home/bls/Downloads/code/bbot/bbot/modules/internal/excavate.py:392> cb=[Task.task_wakeup()] created at /usr/lib/python3.11/asyncio/tasks.py:384> took 0.274 seconds
```

### PR DESCRIPTION
**Pull request by liquidsec**
- [x] Rework Spider Danger / Spider Max
- [x] Get all old tests passing
- [x] Write tests for new functionality
- [x] Documentation for Yara features
- [ ] Clean up / refactoring

### CODE REVIEW

File: bbot/core/event/base.py [hunk 1]
```diff
@@ -1124,6 +1147,40 @@ class URL_HINT(URL_UNVERIFIED):
...
+        # this is a hack which needs to be replaced with a real fix in bbot-2.0 branch
+        if self.scan is not None:
+            if self.scan.config.get("url_querystring_remove", True) == False:
+                from urllib.parse import urlparse
+
+                parsed_url = urlparse(url)
+
+                if self.scan.config.get("url_querystring_collapse", True) == True:
+                    # Only consider parameter names in dedup (collapse values)
+                    cleaned_query = "|".join(sorted([p.split("=")[0] for p in parsed_url.query.split("&")]))
```
Comment (TheTechromancer) [inline bbot/core/event/base.py]: We should avoid manually parsing query strings and use `parse_qs()` instead.

File: bbot/modules/internal/excavate.py [hunk 2]
```diff
@@ -1,476 +1,858 @@
...
-            context=f"{{module}}'s CSP extractor found {{event.type}}: {{event.data}} in CSP header",
-        )
-
-
-class HostnameExtractor(BaseExtractor):
-    regexes = {}
+        self.name = ""
+
+    async def preprocess(self, r, event, discovery_context):
+        self.discovery_context = discovery_context
```
Comment (TheTechromancer) [inline bbot/modules/internal/excavate.py]: Since the same rule might be used to process multiple events concurrently, we need to avoid setting event-specific data on `self`.

File: bbot/modules/internal/excavate.py [hunk 3]
```diff
@@ -1,476 +1,858 @@
...
+
+        for identifier, results in yara_results.items():
+            for result in results:
+                event_data = {"host": str(event.host), "url": event.data.get("url", "")}
+                event_data["description"] = f"{self.discovery_context} {yara_rule_settings.description}"
+                if yara_rule_settings.emit_match:
+                    event_data["description"] += f" [{result}]"
+                await self.report(event_data, event, yara_rule_settings)
+
+    async def report_prep(self, event_data, event_type, event, tags):
```
Comment (TheTechromancer) [inline bbot/modules/internal/excavate.py:181]: Each of the functions on the base `ExcavateRule` class need to have docstrings that explain their purpose and how they fit with the other functions (when/why to override, etc.).

File: bbot/modules/internal/excavate.py [hunk 4]
```diff
@@ -1,476 +1,858 @@
...
+                parsed_url.scheme,
+                parsed_url.netloc,
+                parsed_url.path,
+                "",
+                querystring if self.retain_querystring else "",
+                "",
+            )
         )
 
+    class ParameterExtractor(ExcavateRule):
```
Comment (TheTechromancer) [inline bbot/modules/internal/excavate.py:222]: Should have a docstring explaining how it's different from the other excavate rules (i.e. why it's nested here instead of with the others)

File: bbot/modules/internal/excavate.py [hunk 5]
```diff
@@ -1,476 +1,858 @@
...
-                        result = path
+    class excavateTestRule(ExcavateRule):
+        yara_rules = {
+            "SearchForText": 'rule SearchForText { meta: description = "Contains the text AAAABBBBCCCC" strings: $text = "AAAABBBBCCCC" condition: $text }',
+            "SearchForText2": 'rule SearchForText2 { meta: description = "Contains the text DDDDEEEEFFFF" strings: $text2 = "DDDDEEEEFFFF" condition: $text2 }',
+        }
+    """
 
-                yield result, name
+    watched_events = ["HTTP_RESPONSE", "RAW_TEXT"]
```
Comment (TheTechromancer) [inline bbot/modules/internal/excavate.py]: Watches for `RAW_TEXT`, but it seems like `handle_event` is only set up to handle `HTTP_RESPONSE`. We should write tests for `RAW_TEXT` (or just remove it for now).

File: bbot/test/test_step_1/test_manager_scope_accuracy.py [hunk 6]
```diff
@@ -340,6 +340,9 @@ def custom_setup(scan):
     assert 1 == len([e for e in events if e.type == "IP_ADDRESS" and e.data == "127.0.0.77" and e.internal == False and e.scope_distance == 1])
     assert 0 == len([e for e in events if e.type == "OPEN_TCP_PORT" and e.data == "127.0.0.77:8888"])
 
+
+    for e in all_events:
```
Comment (TheTechromancer) [inline bbot/test/test_step_1/test_manager_scope_accuracy.py]: let's not forget to remove this

File: bbot/test/test_step_2/module_tests/test_module_discord.py [hunk 7]
```diff
@@ -4,15 +4,15 @@
...
     modules_overrides = ["discord", "excavate", "badsecrets", "httpx"]
 
     webhook_url = "https://discord.com/api/webhooks/1234/deadbeef-P-uF-asdf"
     config_overrides = {"modules": {"discord": {"webhook_url": webhook_url}}}
 
     def custom_setup(self, module_test):
         respond_args = {
-            "response_data": '<html><body><p>Express Cookie Test<a href="ftp://127.0.0.1/asdf.txt"/></p></body></html>',
+            "response_data": '<html><body><p>Express Cookie Test<a href="ftp://asdf.foo.bar/asdf.txt"/></p></body></html>',
             "headers": {
```
Comment (TheTechromancer) [inline bbot/test/test_step_2/module_tests/test_module_discord.py:16]: Why was this necessary? Are we unable to excavate URLs with IPs?

Comment (TheTechromancer) [reply bbot/test/test_step_2/module_tests/test_module_discord.py:16]: URL excavation with IPs is a must-have feature before we merge.

File: docs/modules/internal_modules.md [hunk 8]
```diff
@@ -0,0 +1,85 @@
+# List of Modules
+
+## What are internal modules?
+
+Internal modules just like regular modules, except that they run all the time. They do not have to be explicitly enabled. They can, however, be explicitly disabled if needed.
```
Comment (TheTechromancer) [inline docs/modules/internal_modules.md]: Missing a word somewhere in there

File: docs/modules/internal_modules.md [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+speculate: True
+# Passively search event data for URLs, hostnames, emails, etc.
+excavate: True
+# Summarize activity at the end of a scan
+aggregate: True
+# DNS resolution
+dnsresolve: True
+# Cloud provider tagging
+cloudcheck: True
+```
```
Comment (TheTechromancer) [inline docs/modules/internal_modules.md:20]: Todo: fix these before merge

File: bbot/core/event/base.py (hunk 1 above)
Comment (liquidsec) [reply bbot/core/event/base.py]: https://github.com/blacklanternsecurity/bbot/pull/1528

File: bbot/modules/internal/excavate.py (hunk 2 above)
Comment (liquidsec) [reply bbot/modules/internal/excavate.py]: https://github.com/blacklanternsecurity/bbot/pull/1471/commits/b191770fcb157ac1b7043881ba0b83edf77b56b2

File: bbot/test/test_step_1/test_manager_scope_accuracy.py (hunk 6 above)
Comment (liquidsec) [reply bbot/test/test_step_1/test_manager_scope_accuracy.py]: https://github.com/blacklanternsecurity/bbot/pull/1471/commits/07d0d50c9fb96d2063cce99cfd963e9435c51ea6

File: docs/modules/internal_modules.md (hunk 9 above)
Comment (liquidsec) [reply docs/modules/internal_modules.md:20]: Whats the issue here?

File: docs/modules/internal_modules.md (hunk 8 above)
Comment (liquidsec) [reply docs/modules/internal_modules.md]: https://github.com/blacklanternsecurity/bbot/pull/1471/commits/76b10814ee4bb7ab6bc5e5cf48a9e9a4773f679c

File: bbot/modules/internal/excavate.py (hunk 3 above)
Comment (liquidsec) [reply bbot/modules/internal/excavate.py:181]: https://github.com/blacklanternsecurity/bbot/pull/1471/commits/7560cc8fc393d58fb397aad2b13ff83d51745d60
https://github.com/blacklanternsecurity/bbot/pull/1471/commits/5ded168f39c61e284ecb2620ad0b2cfe6aa300b2

File: bbot/modules/internal/excavate.py (hunk 5 above)
Comment (liquidsec) [reply bbot/modules/internal/excavate.py]: I think removing for this PR is the best option. We can add it back in as a complete package, with new tests, etc.
https://github.com/blacklanternsecurity/bbot/pull/1471/commits/661331ff641d65e96cdf1d5379b08bb7402c1746

File: bbot/test/test_step_2/module_tests/test_module_discord.py (hunk 7 above)
Comment (liquidsec) [reply bbot/test/test_step_2/module_tests/test_module_discord.py:16]: IP urls work now (as in, they always worked). I honestly can't remember why i made that change in the tests at this point. But i created a new test that explicitly verifies it:

https://github.com/blacklanternsecurity/bbot/pull/1471/commits/3e52d624e44dacc6f9faa4956173db573134367d

File: docs/modules/internal_modules.md (hunk 9 above)
Comment (TheTechromancer) [reply docs/modules/internal_modules.md:20]: Some of the module names aren't right but that's my fault; I'll handle changing them.
